util/authkit/src/index.ts

```diff
@@ -1 +1,224 @@
+    displayName: string;
+    roles: string[];
+};
+
+/** What the login check gets passed: the data from the login form. */
+export type LoginData = {
+    userid: string;
+    password: string;
+};
+
+/** Function that contains the actual login logic. */
+export type LoginCheck = (data: LoginData) => Promise<LoginOutcome>;
+
+/** Specifies either a host+port or a Unix socket to listen on/connect to. */
+export type HttpLocation =
+    | { host: string; port: number; }
+    | { socketPath: string; }
+
+/** Options for the login handler server. */
+export type ServerOptions = {
+    /**
+     * Function to check whether a login attempt is successful, i.e. the actual
+     * login logic you have to provide.
+     */
+    check: LoginCheck;
+
+    /** What the server should listen on. */
+    listen: HttpLocation,
+
+    /**
+     * How to reach Tobira. On successful logins, `POST /~session` is sent to
+     * the specified Tobira. Default: `{ host: "localhost", port: 3080 }`.
+     */
+    tobira?: HttpLocation;
+
+    /** Called once the server starts listening (e.g. the port is open). */
+    onListen?: () => void;
+};
+
+
+/**
+ * Starts an HTTP server that handles Tobira logins by calling the specified
+ * `check` function.
+ *
+ * While this is only intended for login requests (`POST /~login`), the path of
+ * the request is actually completely ignored. This might be convenient for
+ * some situations and does not introduce any problems. However, some other
+ * conditions are checked:
+ * - If method is not POST => 405 Method not allowed
+ * - If the "Content-Type" header is missing or does not start
+ *   with "application/x-www-form-urlencoded" => 400 Bad request
+ * - If the body is not UTF-8 => 400 Bad request
+ * - If the body is not a correctly encoded query string with `userid` and
+ *   `password` fields exactly once => 400 Bad request
+ *
+ * If all those checks pass, the `check` function is called with the login data.
+ * Depending on what it returns, either "Forbidden" is replied or a user
+ * session in Tobira is created and an appropriate response is sent.
+ */
+export const runServer = async (options: ServerOptions): Promise<void> => (
+    new Promise((resolve, reject) => {
+        const server = http.createServer((req, res) => {
+            listener(req, res, options)
+                .then(() => res.end())
+                .catch(e => {
+                    if (e instanceof ErrorResponse) {
+                        res.writeHead(e.code);
+                        if (e.body) {
+                            res.write(e.body);
+                        }
+                    } else {
+                        res.writeHead(StatusCode.INTERNAL_SERVER_ERROR);
+                        res.write("Internal server error");
+                    }
+                    res.end();
+                });
+        });
+        server.on("listening", () => options.onListen?.());
+        server.on("error", e => reject(e));
```

Comment:
Not looking at code at all/too closely is haaaard! xP

```suggestion
        server.on("error", reject);
```

and something like this happens quite often in this PR.